File: day14/sol2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <array>
#include <string>
#include <sstream>
#include <regex>
#include <numeric>
#include <cmath>

using namespace std;

struct Vector2i {
    int x, y;

    Vector2i() : x(0), y(0) {}
    Vector2i(int x, int y) : x(x), y(y) {}

    Vector2i operator+(const Vector2i& other) const {
        return Vector2i(x + other.x, y + other.y);
    }

    Vector2i operator*(int scalar) const {
        return Vector2i(x * scalar, y * scalar);
    }

    bool operator==(const Vector2i& other) const {
        return x == other.x && y == other.y;
    }

    Vector2i& operator%=(const Vector2i& dims) {
        x %= dims.x;
        y %= dims.y;
        return *this;
    }

    bool all_less_than(const Vector2i& other) const {
        return x < other.x && y < other.y;
    }

    bool any_greater_than(const Vector2i& other) const {
        return x > other.x || y > other.y;
    }
};

void log(const string& file_name, const string& smth) {
    if (file_name.find("test.in") != string::npos) {
        cout << smth << endl;
    }
}

vector<pair<Vector2i, Vector2i>> load_data(const string& file_name) {
    ifstream file(file_name);
    string line;
    vector<pair<Vector2i, Vector2i>> result;

    regex pattern("-?\\d+");
    while (getline(file, line)) {
        smatch match;
        auto begin = sregex_iterator(line.begin(), line.end(), pattern);
        auto end = sregex_iterator();

        vector<int> numbers;
        for (auto it = begin; it != end; ++it) {
            numbers.push_back(stoi(it->str()));
        }

        Vector2i position(numbers[0], numbers[1]);
        Vector2i velocity(numbers[2], numbers[3]);
        result.emplace_back(position, velocity);
    }

    return result;
}

int get_quadrant(const Vector2i& pos, const Vector2i& dims) {
    array<Vector2i, 4> quads = {
        Vector2i(dims.x / 2, dims.y / 2),
        Vector2i(dims.x, dims.y / 2),
        Vector2i(dims.x / 2, dims.y),
        Vector2i(dims.x, dims.y)
    };

    if (pos.all_less_than(quads[0])) {
        return 0;
    } else if (pos.any_greater_than(quads[0]) && pos.all_less_than(quads[1])) {
        return 1;
    } else if (pos.any_greater_than(quads[0]) && pos.all_less_than(quads[2])) {
        return 2;
    } else if (pos.x > quads[2].x && pos.y > quads[1].y) {
        return 3;
    }

    return -1;
}

void sol1(const string& file_name, const Vector2i& dims) {
    auto robots = load_data(file_name);

    for (int i = 1; i < 40000; ++i) {
        array<int, 4> quad_cnt = {0, 0, 0, 0};
        vector<Vector2i> positions;

        for (const auto& [p, v] : robots) {
            Vector2i pos = p + v * i;
            pos %= dims;
            positions.push_back(pos);

            int quad = get_quadrant(pos, dims);
            if (quad != -1) {
                quad_cnt[quad]++;
            }
        }

        if (file_name.find("input") != string::npos) {
          std::cout << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! " << i << std::endl;
            for (int r = 0; r < dims.y; ++r) {
                for (int c = 0; c < dims.x; ++c) {
                    int count = count_if(positions.begin(), positions.end(), [&](const Vector2i& pos) {
                        return pos == Vector2i(c, r);
                    });
                    cout << (count > 0 ? "#" : ".") << "";
                }
                cout << endl;
            }
            cout << endl;
        }
    }
}

int main() {
    cout << "Star 1:" << endl;
    cout << "test: ";
    sol1("test.in", Vector2i(11, 7));
    cout << "answer: ";
    sol1("input.in", Vector2i(101, 103));

    return 0;
}

File: day14/sol2.py
from collections import defaultdict
import regex as re
import sys
import numpy as np
import math

def log(file_name, smth):
  if "test.in" in file_name:
    print(smth)

def load_data(file_name):
  with open(file_name) as f:
    data = f.readlines()
  
  res = [] 
  for line in data:
    p, v = line.split()
    pattern = r"-?\d+"
    p = re.findall(pattern, p)
    v = re.findall(pattern, v)
    res.append((np.array(list(map(int, p))), np.array(list(map(int, v)))))

  return res

def get_quadrant(pos, dims):
  quads = [[dims[0] // 2, dims[1] // 2], [dims[0], dims[1] // 2], [dims[0] // 2, dims[1]], [dims[0], dims[1]]]
  quads = [np.array(x) for x in quads]
  if np.all(pos < quads[0]):
    return 0
  elif np.any(pos > quads[0]) and np.all(pos < quads[1]):
    return 1
  elif np.any(pos > quads[0]) and np.all(pos < quads[2]):
    return 2
  elif pos[0] > quads[2][0] and pos[1] > quads[1][1]:
    return 3
  
  return -1

def sol1(file_name, dims):
  robots = load_data(file_name)
  
  quad_cnt = [0, 0, 0, 0]
  positions = []
  for p, v in robots:
    pos = p + v * 100
    pos[0] %= dims[0]
    pos[1] %= dims[1]
    positions.append(pos)
    quad = get_quadrant(pos, dims)
    if quad != -1:
      quad_cnt[quad] += 1 
    
  if "test" in file_name:
    for r in range(dims[1]):
      for c in range(dims[0]):
        cnt = sum(np.array_equal(pos, [c, r]) for pos in positions)
        print(cnt if cnt > 0 else '.', end='')
      print()
   
  print(quad_cnt)
  return math.prod(quad_cnt)
  
print(f"Star 1:\n test: {sol1('test.in', np.array([11, 7], dtype=int))}\n answer: {sol1('input.in', np.array([101, 103], dtype=int))}\n")